libraries/standard/mqtt/src/mqtt.c

```diff
@@ -130,6 +234,447 @@ static int32_t sendPacket( MQTTContext_t * pContext,
+        status = MQTTNoDataAvailable;
+    }
+
+    return status;
+}
+
+/*-----------------------------------------------------------*/
+
+static MQTTStatus_t receivePacket( MQTTContext_t * const pContext,
+                                   MQTTPacketInfo_t incomingPacket,
+                                   uint32_t remainingTimeMs )
+{
+    MQTTStatus_t status = MQTTSuccess;
+    int32_t bytesReceived = 0;
+    size_t bytesToReceive = 0U;
+
+    assert( pContext != NULL );
+
+    if( incomingPacket.remainingLength > pContext->networkBuffer.size )
+    {
+        LogErrorWithArgs( "Incoming packet length %u exceeds network buffer size %u."
+                          "Incoming packet will be dumped.",
+                          incomingPacket.remainingLength,
+                          pContext->networkBuffer );
+        status = discardPacket( pContext,
+                                incomingPacket.remainingLength,
+                                remainingTimeMs );
+    }
+    else
+    {
+        bytesToReceive = incomingPacket.remainingLength;
+        bytesReceived = recvExact( pContext, bytesToReceive, remainingTimeMs );
+        if( bytesReceived == ( int32_t ) bytesToReceive )
+        {
+            /* Receive successful, bytesReceived == bytesToReceive. */
+            LogInfoWithArgs( "Packet received. ReceivedBytes=%d.",
+                             bytesReceived );
+        }
+        else
+        {
+            LogErrorWithArgs( "Packet reception failed. ReceivedBytes=%d, "
+                              "ExpectedBytes=%u.",
+                              bytesReceived,
+                              bytesToReceive );
+            status = MQTTRecvFailed;
+        }
+        
+    }
+
+    return status;
+}
+
+/*-----------------------------------------------------------*/
+
+static MQTTStatus_t sendPublishAcks( MQTTContext_t * const pContext,
+                                     uint16_t packetId,
+                                     MQTTPublishState_t * pPublishState )
+{
+    MQTTStatus_t status = MQTTSuccess;
+    MQTTPublishState_t newState = MQTTStateNull;
+    int32_t bytesSent = 0;
+    uint8_t packetTypeByte = 0U;
+    MQTTPubAckType_t packetType;
+
+    assert( pContext != NULL );
+    assert( pPublishState != NULL );
+
+    switch( *pPublishState )
+    {
+        case MQTTPubAckSend:
+            packetTypeByte = MQTT_PACKET_TYPE_PUBACK;
+            packetType = MQTTPuback;
+            break;
+        case MQTTPubRecSend:
+            packetTypeByte = MQTT_PACKET_TYPE_PUBREC;
+            packetType = MQTTPubrec;
+            break;
+        case MQTTPubRelSend:
+            packetTypeByte = MQTT_PACKET_TYPE_PUBREL;
+            packetType = MQTTPubrel;
+            break;
+        case MQTTPubCompSend:
+            packetTypeByte = MQTT_PACKET_TYPE_PUBCOMP;
+            packetType = MQTTPubcomp;
+            break;
+        default:
+            /* Nothing to send. */
```

Comment:
Is this expected to be possible in normal operation?  If not, should we log an error?  If so, I'd add elaborate this comment with more detail, e.g. _"This function is called for all possible states, take no action on these."_

Reply:
Yes, it is called for all possible states, but action is only required for states indicating an ack should be sent. I will update the comment.